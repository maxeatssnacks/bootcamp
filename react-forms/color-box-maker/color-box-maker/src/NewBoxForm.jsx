import { useState } from "react";

const NewBoxForm = ({ createBox }) => {
    const INITIAL_STATE = { id: '', color: "#ffffff", width: 100, height: 100 };
    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleSubmit = evt => {
        evt.preventDefault();
        createBox({ ...formData });
        setFormData(INITIAL_STATE);

    }

    const handleChange = evt => {
        const { name, value } = evt.target;
        setFormData(formData => ({
            ...formData,
            [name]: name === 'width' || name === 'height' ? parseInt(value) : value
        }))
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="color">Color: </label>
            <input id="color" name="color" type="color" value={formData.color} onChange={handleChange} />
            <label htmlFor="width">Width: </label>
            <input id="width" name="width" type="number" value={formData.width} onChange={handleChange} />
            <label htmlFor="height">Height: </label>
            <input id="height" name="height" type="number" value={formData.height} onChange={handleChange} />
            <button type="submit">Create a new box!</button>
        </form>
    )
}

export default NewBoxForm;