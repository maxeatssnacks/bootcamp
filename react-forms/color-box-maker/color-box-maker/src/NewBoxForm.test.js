import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

describe('NewBoxForm', () => {
    // Smoke test
    test('renders without crashing', () => {
        render(<NewBoxForm createBox={() => { }} />);
    });

    // Snapshot test
    test('matches snapshot', () => {
        const { asFragment } = render(<NewBoxForm createBox={() => { }} />);
        expect(asFragment()).toMatchSnapshot();
    });

    // Critical business logic test
    test('form submits correctly', () => {
        const mockCreateBox = jest.fn();
        render(<NewBoxForm createBox={mockCreateBox} />);

        // Find form inputs and submit button
        const colorInput = screen.getByLabelText('Color:');
        const widthInput = screen.getByLabelText('Width:');
        const heightInput = screen.getByLabelText('Height:');
        const submitButton = screen.getByText('Create a new box!');

        // Fill out the form
        fireEvent.change(colorInput, { target: { value: '#00FF00' } });
        fireEvent.change(widthInput, { target: { value: '200' } });
        fireEvent.change(heightInput, { target: { value: '150' } });

        // Submit the form
        fireEvent.click(submitButton);

        // Check if createBox was called with the correct arguments
        expect(mockCreateBox).toHaveBeenCalledWith(expect.objectContaining({
            id: '',
            color: expect.stringMatching(/^#00ff00$/i),  // case-insensitive match
            width: 200,
            height: 150
        }));

        // Check if form fields are reset after submission
        expect(colorInput.value.toLowerCase()).toBe('#ffffff');
        expect(widthInput.value).toBe('100');
        expect(heightInput.value).toBe('100');
    });
});