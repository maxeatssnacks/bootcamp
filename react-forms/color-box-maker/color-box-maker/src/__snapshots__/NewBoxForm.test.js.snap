// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewBoxForm matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color: 
    </label>
    <input
      id="color"
      name="color"
      type="color"
      value="#ffffff"
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value="100"
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value="100"
    />
    <button
      type="submit"
    >
      Create a new box!
    </button>
  </form>
</DocumentFragment>
`;
